shell/linux/Makefile

```diff
@@ -11,47 +9,134 @@ USE_OSS := 1
 #USE_PULSEAUDIO := 1
 USE_X11 := 1
 
+CXX=${CC_PREFIX}g++
+CC=${CC_PREFIX}gcc
+AS=${CC_PREFIX}as
+STRIP=${CC_PREFIX}strip
+LD=${CC}
+
 MFLAGS := 
 ASFLAGS := 
 LDFLAGS :=
+INCS :=
+LIBS :=
 CFLAGS := 
 CXXFLAGS :=
 
-ifeq (,$(ARCH))
+# Platform auto-detection
+# Can be overridden by using:
+#   make platform=x64
+ifeq (,$(platform))
     ARCH = $(shell uname -m)
+    ifeq ($(ARCH), $(filter $(ARCH), i386 i686))
+        platform = x86
+    else ifeq ($(ARCH), $(filter $(ARCH), x86_64 AMD64))
+        platform = x64
+    else ifneq (,$(findstring arm,$(ARCH)))
+        HARDWARE = $(shell grep Hardware /proc/cpuinfo)
+        ifneq (,$(findstring BCM2709,$(HARDWARE)))
+            platform = rpi2
+        else ifneq (,$(findstring AM33XX,$(HARDWARE)))
+            platform = beagle
+        else ifneq (,$(findstring Pandora,$(HARDWARE)))
+            platform = pandora
+        else
+            $(error Unsupported Hardware)
+        endif
+    else ifneq (,$(findstring mips,$(ARCH)))
+        platform = gcwz
+    else
+        $(error Unsupported CPU architecture)
+    endif
 endif
 
-$(info Building arch: $(ARCH))
-ifeq ($(ARCH), $(filter $(ARCH), i386 i686))
+$(info Platform: $(platform))
+
+# Generic 32 bit x86 (a.k.a. i386/i486/i686)
+ifneq (,$(findstring x86,$(platform)))
     X86_REC := 1
+    NOT_ARM := 1
     MFLAGS += -m32
     ASFLAGS += -32
     LDFLAGS += -m32
-    CFLAGS += -m32 -D TARGET_LINUX_x86
+    CFLAGS += -m32 -D TARGET_LINUX_x86 -D TARGET_NO_AREC -fsingle-precision-constant
     CXXFLAGS += -fno-exceptions
-else
+
+# Generic 64 bit x86 (a.k.a. x64/AMD64/x86_64/Intel64/EM64T)
+else ifneq (,$(findstring x64,$(platform)))
     X64_REC := 1
-    CFLAGS += -D TARGET_LINUX_x64
+    NOT_ARM := 1
+    CFLAGS += -D TARGET_LINUX_x64 -D TARGET_NO_AREC -fsingle-precision-constant
     CXXFLAGS += -fexceptions
+
+# LinCPP
+else ifneq (,$(findstring lincpp,$(platform)))
+    CPP_REC := 1
+    NOT_ARM := 1
+    CFLAGS += -D TARGET_LINUX_x64 -D TARGET_NO_JIT
+    CXXFLAGS += -fexceptions -std=gnu++11
```

Comment:
Do the other implementation  support `-std=gnu++11` as well? If so, I could apply it globally, not just for LinCPP.